### PR TITLE
[DASHTree] stop waiting when new period appears to fix stalls during playback

### LINKED ISSUE #1540: MPD with new period gets stuck during playback
## Bug report
Regression of #1175 

### Describe the bug
Previously would show "Switching to new Period"

Now playback gets stuck and the logs show mpd refresh over and over without switching over to the next period

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

## Actual Behavior

```
[AS-0] Download finished: 1711567398226item-06item_Segment-664560.mp4 (downloaded 2285902 byte, speed 57610955.24 byte/s)
[AS-1] Download finished: 1711567398226item-07item_Segment-664560.mp4 (downloaded 108040 byte, speed 3620498.81 byte/s)
ensureSegment: [AS-0] Begin WaitForSegment stream rep. id "1711567398226item-06item" period id "5760"
Download finished: master.mpd (downloaded 21646 byte, speed 322131.44 byte/s)
RefreshLiveSegments: MPD update - Done (repr. id "1711567398226item-06item", period id "5760")
RefreshLiveSegments: MPD update - Done (repr. id "1711567398226item-07item", period id "5760")
RefreshLiveSegments: Inserting new Period (id=5761, start=2593986796)
Download finished: master.mpd (downloaded 21826 byte, speed 146706.51 byte/s)
RefreshLiveSegments: MPD update - Done (repr. id "1711567398226item-06item", period id "5760")
RefreshLiveSegments: MPD update - Done (repr. id "1711567398226item-07item", period id "5760")
Download finished: master.mpd (downloaded 21844 byte, speed 713352.29 byte/s)
```

### To Reproduce

1. 
2.
3.

### Debuglog

The debuglog can be found here:

### MPD/M3U8s/ISM

An example or copy of a manifest (or manifests for HLS - master and variants) can be found here:

### Screenshots 
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [ ] Linux
 - [ ] OSX
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name:
 - Kodi version:
 - inputstream.adaptive version:

*note: Once the issue is made we require you to update it with new information should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

### PR DESCRIPTION
**Pull request by tmm1**
## Description

fixes https://github.com/xbmc/inputstream.adaptive/issues/1540
cc #1175 

## How has this been tested?

tested on stream and confirmed no more stalls

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the Wiki documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: src/parser/DASHTree.cpp [hunk 1]
```diff
@@ -1691,6 +1691,11 @@ void adaptive::CDashTree::RefreshLiveSegments()
                   LOG::LogF(LOGDEBUG,
                             "MPD update - No new segments (repr. id \"%s\", period id \"%s\")",
                             repr->GetId().data(), period->GetId().data());
+                  if (repr->IsWaitForSegment() &&
+                      period->GetId() != m_periods.back()->GetId())
+                  {
+                    repr->SetIsWaitForSegment(false);
+                  }
```
Comment (CastagnaIT) [inline src/parser/DASHTree.cpp:1698]: you havent provide any explanation nor manifests that explain why this is needed
when there are no segments you must not remove the "wait for segments"
the demuxer callbacks must be kept stopped

File: src/parser/DASHTree.cpp [hunk 2]
```diff
@@ -1737,7 +1742,9 @@ void adaptive::CDashTree::RefreshLiveSegments()
                 }
               }
 
-              if (repr->IsWaitForSegment() && (repr->get_next_segment(repr->current_segment_)))
+              if (repr->IsWaitForSegment() &&
+                  (repr->get_next_segment(repr->current_segment_) ||
+                   period->GetId() != m_periods.back()->GetId()))
```
Comment (CastagnaIT) [inline src/parser/DASHTree.cpp:1747]: this addition is not clear
`period->GetId() != m_periods.back()->GetId()`

**Review by CastagnaIT — CHANGES_REQUESTED** [review]
i would have a way to replicate this, since the changes are not clear
and the root of cause may be elsewhere

if a new period appears, it should trigger the last segment check on AdaptiveStream when the current one is ending
if this dont happens can suggest some other problem, for example is missing to be updated the period duration

**Comment by CastagnaIT** [comment]
closed since atm these changes are not so clear
if you are willing to provide copies of the mpd's downloaded by ISA during playback with related full debug log,
will be possible for us to understand the situation and then make a fix or allow you to modify the code appropriately
i will keep open related issue for about a week, if no responses also the issue will be closed

**Comment by tmm1** [comment]
Was discussed in detail during initial implementation. https://teamkodi.slack.com/archives/C0U5RMJSD/p1677448422522129

**Comment by CastagnaIT** [comment]
on that discussion you are talking about test stream
https://livesim.dashif.org/livesim2/periods_20/testpic_2s/Manifest.mpd

this one has been already tested on Omega over last recent changes,
anyway testing it on current master, results with a correct playback behaviour:

https://paste.kodi.tv/uhejasiquq.kodi

[manifest_1716875099.txt](https://github.com/xbmc/inputstream.adaptive/files/15464717/manifest_1716875099.txt)
[manifest_1716875101.txt](https://github.com/xbmc/inputstream.adaptive/files/15464718/manifest_1716875101.txt)

> Inserting new Period (id=P9538195, start=1716875100000)
> ...
> Check for last segment (period end PTS: 1716875100000, segment end PTS: 1716875100000)
> ...
> Switching to new Period (id=P9538195, start=1716875100000, seq=2)

and segment downloaded from switched period starts with the right numeration

> [AS-2] Download finished: https://livesim.dashif.org/livesim2/periods_20/testpic_2s/A48/858437550.m4s (downloaded 13188 byte, speed 107627.00 byte/s)
> [AS-3] Download finished: https://livesim.dashif.org/livesim2/periods_20/testpic_2s/V300/858437550.m4s (downloaded 38009 byte, speed 324261.00 byte/s)

so your live stream looks to be different than this test stream

**Pull request closed without merging**